### PR TITLE
Database integration

### PR DESCRIPTION
**Pull request by LhermannSauer**
roastMe();

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -1,18 +1,17 @@
-import express from "express";
-import "dotenv/config";
+import express from 'express';
+import 'reflect-metadata';
 
-// init import
-import { initMiddleware } from "./server/src/init/middleware";
+import * as dotenv from 'dotenv';
+dotenv.config({ path: `./.env.${process.env.NODE_ENV}` });
```
Comment (titogithub) [inline app.ts]: should we check if this file exist ?

File: server/src/common/errors.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export class NotFoundError extends Error {
```
Comment (titogithub) [inline server/src/common/errors.ts:1]: mmm I don't think it's worth splitting this kind of errors. Instead we can group it like
```
class ErrorResponseDTO extends Error {
message: string;
statusCode: number;
reason: any
 constructor(...) { ... }
}
```
not found could be `new ErrorResponseDTO(`The ${entity} with the given ID was not found`, StatusCodes.NOT_FOUND, 'if needed any detail')`

invalid parameter error could be `new ErrorResponseDTO('Bad request', StatusCodes.BAD_REQUEST, 'if needed any detail')`

File: server/src/controller/todoItemController.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  priority: number;
+}
+
+export class TodoItemController {
+  static getItems = async (): Promise<TodoItem[]> => {
+    return await todoItemRepo.find();
+  };
+
+  static getItemById = async (id: number): Promise<TodoItem> => {
+    const item = await todoItemRepo.findOneBy({ id });
```
Comment (titogithub) [inline server/src/controller/todoItemController.ts:27]: Usually the controller validates the request and parses the response. Id is no being validated. The controller layer call the services layer. The service layer calls the repository layer.

File: server/src/middleware/error.middleware.ts [hunk 4]
```diff
@@ -1,7 +1,12 @@
-import { Express, NextFunction, Request, Response } from "express";
+import { Express, NextFunction, Request, Response } from 'express';
+import { InvalidParameterError, NotFoundError } from '../common/errors';
 
 export function errorMiddleware(app: Express): void {
   app.use((error: any, req: Request, res: Response, next: NextFunction) => {
-    res.status(500).json({ error: "Internal Server Error" });
+    if (error instanceof NotFoundError) {
```
Comment (titogithub) [inline server/src/middleware/error.middleware.ts:6]: consider not using `else if` syntax

**Review by titogithub — APPROVED** [review]
Concepts to be revisited. 
Inversion of control, dependency injection. I think the best way of understanding how useful those concepts are, is trying to add test to this API.
Consider adding any data validation library.

File: app.ts (hunk 1 above)
Comment (LhermannSauer) [reply app.ts]: Yes we should

File: server/src/controller/todoItemController.ts (hunk 3 above)
Comment (LhermannSauer) [reply server/src/controller/todoItemController.ts:27]: Added ID validation

File: server/src/middleware/error.middleware.ts (hunk 4 above)
Comment (LhermannSauer) [reply server/src/middleware/error.middleware.ts:6]: Refactored this to avoid using else if.

File: server/src/controller/todoItemController.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    return item;
+  };
+
+  static removeItemById = async (id: number): Promise<boolean> => {
+    if (!Number.isInteger(id) || id <= 0) throw new InvalidParameterError('ID');
+
+    const result = await todoItemRepo.delete({ id });
+
+    return typeof result.affected === 'number';
```
Comment (titogithub) [inline server/src/controller/todoItemController.ts:58]: `return result.affected > 0`

**Pull request merged**